### PR TITLE
[transforms] Add ConstantFolding as transform

### PR DESCRIPTION
**Pull request by kshitij12345**
Related #1250 

Add ConstantFolding as a transform. It is added as `transform_traces_pre_prologue` so that successive passes can use the simplified compute. We use PyTorch eager to evaluate on constant values see `compute_with_concrete_tensor_and_pytorch_eager`.

NOTE - This is currently opt-in. I think we should figure out if this should be in default and how that would look in a seperate PR.

Misc Changes
* Few fixes to `get_symbols_to_last_used_variables` from utils.py
* Update signature of `thunder.torch.add` to match the PyTorch [torch.add](https://pytorch.org/docs/stable/generated/torch.add.html) and torch.sub signature (alpha should be a number and not None).

Example Traces

Function
```python
def forward(x):
    scale_t = torch.tensor([2.])
    getitem = scale_t[0]
    getitem_2 = torch.tensor([2., 3.])[0]
    return x + getitem + getitem_2

```

Original Computation Trace
```python
def computation(x):
  # x: "cpu f32[3, 3]"
  scale_t = torch.tensor([2.0], device=torch.device("cpu"), dtype=None)  # scale_t: "cpu f32[1]"
    # scale_t = ltorch.tensor([2.0], device=torch.device("cpu"), dtype=None, requires_grad=False, pin_memory=False)  # scale_t: "cpu f32[1]"
      # scale_t = prims.tensor_from_sequence([2.0], dtype=None, device=devices.Device("cpu"))  # scale_t: "cpu f32[1]"
  t3 = torch.tensor([2.0, 3.0], device=torch.device("cpu"), dtype=None)  # t3: "cpu f32[2]"
    # t3 = ltorch.tensor([2.0, 3.0], device=torch.device("cpu"), dtype=None, requires_grad=False, pin_memory=False)  # t3: "cpu f32[2]"
      # t3 = prims.tensor_from_sequence([2.0, 3.0], dtype=None, device=devices.Device("cpu"))  # t3: "cpu f32[2]"
  getitem = Tensor.__getitem__(scale_t, 0)  # getitem: "cpu f32[]"
    # getitem = ltorch.getitem(scale_t, 0)  # getitem: "cpu f32[]"
      # (_,) = prims.shape(scale_t)
      # t17 = prims.slice_prim(scale_t, [0], [1], [1])  # t17: "cpu f32[1]"
      # getitem = prims.squeeze(t17, (0,))  # getitem: "cpu f32[]"
  del scale_t
  getitem_2 = Tensor.__getitem__(t3, 0)  # getitem_2: "cpu f32[]"
    # getitem_2 = ltorch.getitem(t3, 0)  # getitem_2: "cpu f32[]"
      # (_,) = prims.shape(t3)
      # t20 = prims.slice_prim(t3, [0], [1], [1])  # t20: "cpu f32[1]"
      # getitem_2 = prims.squeeze(t20, (0,))  # getitem_2: "cpu f32[]"
  del t3
  t6 = torch.add(x, getitem, alpha=1)  # t6: "cpu f32[3, 3]"
    # t6 = ltorch.add(x, getitem, alpha=1)  # t6: "cpu f32[3, 3]"
      # t6 = prims.add(x, getitem)  # t6: "cpu f32[3, 3]"
  del getitem
  t7 = torch.add(t6, getitem_2, alpha=1)  # t7: "cpu f32[3, 3]"
    # t7 = ltorch.add(t6, getitem_2, alpha=1)  # t7: "cpu f32[3, 3]"
      # t7 = prims.add(t6, getitem_2)  # t7: "cpu f32[3, 3]"
  del t6, getitem_2
  return t7
```

Constant Folded Trace
```python
def computation(x):
  # x: "cpu f32[3, 3]"
  getitem = torch.full((), 2.0, device=torch.device("cpu"), dtype=torch.float32)  # getitem: "cpu f32[]"
    # getitem = ltorch.full((), 2.0, device=torch.device("cpu"), dtype=torch.float32)  # getitem: "cpu f32[]"
      # getitem = prims.full((), 2.0, device=devices.Device("cpu"), dtype=dtypes.float32)  # getitem: "cpu f32[]"
  t6 = torch.add(x, getitem, alpha=1)  # t6: "cpu f32[3, 3]"
    # t6 = ltorch.add(x, getitem, alpha=1)  # t6: "cpu f32[3, 3]"
      # t6 = prims.add(x, getitem)  # t6: "cpu f32[3, 3]"
  t7 = torch.add(t6, getitem, alpha=1)  # t7: "cpu f32[3, 3]"
    # t7 = ltorch.add(t6, getitem, alpha=1)  # t7: "cpu f32[3, 3]"
      # t7 = prims.add(t6, getitem)  # t7: "cpu f32[3, 3]"
  del t6, getitem
  return t7
```

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -1852,9 +1852,9 @@ def silu(a: TensorLike, /, inplace: bool = False) -> TensorLike:
 
 @torchsymbol(torch.add, is_method=True)
 def add(
-    a: NumberLike | TensorLike, b: NumberLike | TensorLike, /, *, alpha: None | Number | TensorLike = None
+    a: NumberLike | TensorLike, b: NumberLike | TensorLike, /, *, alpha: Number | TensorLike = 1
```
Comment (crcrpar) [inline thunder/torch/__init__.py:1855]: :+1:

**Review by crcrpar — COMMENTED** [review]
> Constant Folded Trace
> 
> ```python
> def computation(x):
>   # x: "cpu f32[3, 3]"
>   getitem = torch.full((), 2.0, device=torch.device("cpu"), dtype=torch.float32)  # getitem: "cpu f32[]"
>     # getitem = ltorch.full((), 2.0, device=torch.device("cpu"), dtype=torch.float32)  # getitem: "cpu f32[]"
>       # getitem = prims.full((), 2.0, device=devices.Device("cpu"), dtype=dtypes.float32)  # getitem: "cpu f32[]"
>   t6 = torch.add(x, getitem, alpha=1)  # t6: "cpu f32[3, 3]"
>     # t6 = ltorch.add(x, getitem, alpha=1)  # t6: "cpu f32[3, 3]"
>       # t6 = prims.add(x, getitem)  # t6: "cpu f32[3, 3]"
>   t7 = torch.add(t6, getitem, alpha=1)  # t7: "cpu f32[3, 3]"
>     # t7 = ltorch.add(t6, getitem, alpha=1)  # t7: "cpu f32[3, 3]"
>       # t7 = prims.add(t6, getitem)  # t7: "cpu f32[3, 3]"
>   del t6, getitem
>   return t7
> ```

I'm not quite following why `getitem` is (or needs to be) a tensor.
How would this transform handle CUDA tensor inputs?

**Comment by kshitij12345** [comment]
> I'm not quite following why getitem is (or needs to be) a tensor.
How would this transform handle CUDA tensor inputs?

I think you are confusing `__getitem__` with `torch.Tensor.item`. `__getitem__` always returns a tensor. (I have a feeling the code will fail with call to `torch.Tensor.item`, thanks for the catch).

Eg.
```python
>>> t = torch.randn(3, 3)
>>> t[0, 0]  # getitem
tensor(-0.4925)
>>> t[0, 0].item()
-0.4924606382846832
```

File: thunder/tests/test_transforms.py [hunk 2]
```diff
@@ -497,3 +497,98 @@ def forward(self, x):
...
+    #         # t5 = ltorch.add(x, 5.0, alpha=1)  # t5: "cpu f32[3]"
+    #         # t5 = prims.add(x, 5.0)  # t5: "cpu f32[3]"
+    #     return t5
+
+    # So we check that torch.add has 5.0 in it's arguments.
+    for bsym in exec_trace.bound_symbols:
+        if bsym.sym.id == "add":
+            assert bsym.args[1] == 5.0
+            break
+    else:
```
Comment (t-vi) [inline thunder/tests/test_transforms.py:593]: I don't know. I guess it's OK for the test, but it's one of the things that you have to think a lot what it actually does.

**Review by t-vi — APPROVED** [review]
Seems like a good start, thank you @kshitij12345

**Pull request merged**